src/Nethermind/Nethermind.State/TransientStorageProvider.cs

```diff
@@ -0,0 +1,93 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Resettables;
+using Nethermind.Logging;
+
+namespace Nethermind.State
+{
+    /// <summary>
+    /// EIP-1153 provides a transient store for contracts that doesn't persist
+    /// storage across calls. Reverts will rollback any transient state changes.
+    /// </summary>
+    public class TransientStorageProvider : PartialStorageProviderBase, IPartialStorageProvider
```

Comment:
I don't think TransientStorage needs to be that complicated. It doesn't have the concept of original values, Changes don't need to be tracked as complicated as the tree ones ect.

I would prefer to keep the StorageProvider implementation as it was, and just add a simple Journal<> collection. I think JournalSet<> nor JournalCollection<> won't cut it and we need JournalDictionary<> but it can be implemented in similar  way as JournalSet<>. And this would suffice and keep things simpler. If I am wrong please let me know.

Then this JournalDictionary<> can be used in original StorageProvider Take/Restore snapshot. As well as Commit/Reset for clearing ect.

Reply:
Thanks for the review! If I'm understanding your proposal correctly, I don't think that a JournalDictionary<> would be sufficient since the Transient state also needs to be able to revert to previously stored values at the same index (similar to the PersistentStorageProvider). For example:
1. At depth 0, ContractA performs `TSTORE(x, y) `
2. At depth 1 (recursive call), ContractA performs `TSTORE(x, z)` and then `REVERT`
The previous value `y` should be returned when ContractA calls `TLOAD(x)`. This means that the data structure needs to track all intermediate state similar to the `_intraBlockCache` structure in the original storage provider. Since TransientStorage has so much overlap with PersistentStorage, I tried to minimize code duplication by adding the parent class `PartialStorageProviderBase`. I can move the datastructures (`_intraBlockCache` + `_changes`) to a separate class, but I don't know if it's any better than having it implemented in a parent class if it's not used in other parts of the codebase. 

I do agree that the IPartialStorageProvider interface is awkward since it doesn't properly cover both Persistent and Transient storage. I've removed the interface and cleaned up the irrelevant methods from TransientStorage. LMK if I understood your concerns correctly and if this makes sense.